### DISCUSSION TITLE
using PCM from inside my application code

### DISCUSSION THREAD

**Discussion by abasak24** [opcm/pcm#136] [Q&A]
Hi, 
I am new to PCM and would like to understand some basic things. 

I have a source code app.cpp and two functions inside it (func1 and func2). I want to obtain information such as what is the memory bandwidth utilization during each of the two functions.

Could you please help me understand how to achieve this? What do I need to include in app.cpp to access the PCM functions? Also, where can I learn about what PCM function helps me find the memory bandwidth utilization? Is there a tutorial/manual? 

Thanks!
Abanti

**Comment by ogbrugge** [comment]
Was this solved by now? If not, you can look at how opcm calls the pmu read function twice and calculates the difference, this should also work here for your use case.

**Comment by abasak24** [comment]
Yes, I figured it out. Thanks for responding.

**Comment by yilongli** [comment]
Hi @abasak24, would you mind sharing your code snippet on this? I have been wondering about the same thing. Thank you!

=====================================
**Update**: Never mind, I figured this out from `pcm-memory.cpp` (painfully). Something like the following:
```
#include "pcm/cpucounters.h"

PCM* pcm = PCM::getInstance();
PCM::ErrorCode status = pcm->programServerUncoreMemoryMetrics();
if (status != PCM::Success) return;

// Only measure the memory BW of socket 0.
int socket = 0;
ServerUncorePowerState beforeState = pcm->getServerUncorePowerState(socket);
uint64_t elapsedMicros = pcm->getTickCount(1000000);

// [ Your code here ]

ServerUncorePowerState afterState = pcm->getServerUncorePowerState(socket);
elapsedMicros = pcm->getTickCount(1000000) - elapsedMicros;

double totalReadBytes = 0;
double totalWriteBytes = 0;
for (uint32 c = 0; c < ServerUncorePowerState::maxChannels; c++) {
    totalReadBytes += getMCCounter(c, 0, startState, endState) * 64;
    totalWriteBytes += getMCCounter(c, 1, startState, endState) * 64;
}
std::cout << "Read BW (MB/sec): " << totalReadBytes / elapsedMicros << std::endl;
std::cout << "Write BW (MB/sec): " << totalWriteBytes / elapsedMicros << std::endl;
```

**Comment by varun-jindal** [comment]
Hi all,
It would be really helpful if somebody can point to tutorials for using Intel PCM, I am unable to find any online doc which explains how to use this tool.

**Comment by ogbrugge** [comment]
Hi Varun,

can you be a bit more precise? Do you want info on how to collect performance counters with pcm or do you want info on how to add pcm to your code to measure performance of certain functions/functionality?

**Comment by varun-jindal** [comment]
Hi @ogbrugge ,
I am a newbie and student, I was hoping guidelines to:
1. Setup the software and build necessary files/libraries (eg: to compile code we need "libpcm" and for generating the library the Makefile was in pcm.so folder, which becomes non-intuitive.)
2. Setup the environment and OS permissions required to get the code to compile and run.( Like disabling NMI_Watchdog)
3. Writing and compiling code, probably with few example codes.
If it may help, my situation is that I wanted to use PCM but due to unavailability of documentation (Doxygen reference is really not enough to get started), I switched to PAPI, which has better documentation. I had tried many links which ultimately pointed to Intel Software Manual, but truth is that you can't learn from it as such.
